sched/sched/sched_unlock.c

```diff
@@ -40,278 +40,155 @@
-           * timer for the next time slice.
-           */
-
-          if ((rtcb->flags & TCB_FLAG_POLICY_MASK) == TCB_FLAG_SCHED_RR &&
-              rtcb->timeslice == 0)
-            {
-              /* Yes.. that is the situation.  But one more thing:  The call
-               * to nxsched_merge_pending() above may have actually replaced
-               * the task at the head of the ready-to-run list.  In that
-               * case, we need only to reset the timeslice value back to the
-               * maximum.
-               */
-
-              if (rtcb != this_task())
-                {
-                  rtcb->timeslice = MSEC2TICK(CONFIG_RR_INTERVAL);
-                }
-#ifdef CONFIG_SCHED_TICKLESS
-              else
-                {
-                  nxsched_reassess_timer();
-                }
-#endif
-            }
-#endif
-
-#ifdef CONFIG_SCHED_SPORADIC
-#if CONFIG_RR_INTERVAL > 0
-          else
-#endif
-          /* If (1) the task that was running supported sporadic scheduling
-           * and (2) if its budget slice has already expired, but (3) it
-           * could not slice out because pre-emption was disabled, then we
-           * need to swap the task out now and reassess the interval timer
-           * for the next time slice.
-           */
-
-          if ((rtcb->flags & TCB_FLAG_POLICY_MASK) == TCB_FLAG_SCHED_SPORADIC
-              && rtcb->timeslice < 0)
-            {
-              /* Yes.. that is the situation.  Force the low-priority state
-               * now
-               */
+          sched_preempt_schedule(tcb);
+        }
+    }
+}
 
-              nxsched_sporadic_lowpriority(rtcb);
+/****************************************************************************
+ * Name:  sched_unlock
+ *
+ * Description:
+ *   This function decrements the preemption lock count.  Typically this
+ *   is paired with sched_lock() and concludes a critical section of
+ *   code.  Preemption will not be unlocked until sched_unlock() has
+ *   been called as many times as sched_lock().  When the lockcount is
+ *   decremented to zero, any tasks that were eligible to preempt the
+ *   current task will execute.
+ *
+ ****************************************************************************/
 
-#ifdef CONFIG_SCHED_TICKLESS
-              /* Make sure that the call to nxsched_merge_pending() did not
-               * change the currently active task.
-               */
+void sched_unlock(void)
+{
+  FAR struct tcb_s *tcb = this_task();
 
-              if (rtcb == this_task())
-                {
-                  nxsched_reassess_timer();
-                }
-#endif
-            }
+  if (tcb != NULL)
+    {
+      tcb->lockcount--;
+      DEBUGASSERT(tcb->lockcount >= 0);
+      if (tcb->lockcount == 0)
+        {
+#if (CONFIG_SCHED_CRITMONITOR_MAXTIME_PREEMPTION >= 0) ||\
+    defined(CONFIG_SCHED_INSTRUMENTATION_PREEMPTION)
+          irqstate_t flags = enter_critical_section_wo_note();
```

Comment:
`lockcount` should be protected by `enter_critical_section_wo_note()` if sched note is enabled

Reply:
tcb->lockcount is only accessed by the current task and does not require additional protection

Reply:
To avoid misunderstandings, I have changed the implementation of sched_lock. The lockcount does not need to be protected

Reply:
The changes of csection in `sched_lock()` are what you just updated, right? Please complete the test on your internal products before submitting them to the community. I don't believe that this change has no issues. In SMP, there is no protection between obtaining the TCB and using it.

Reply:
The following situations require entering a critical section when process the TCB:

1 When accessing the TCB of a non-current task.
2 When modifying an attribute of the current TCB, but this attribute may be changed by other tasks.

If you are only accessing or modifying an attribute of the current task's TCB
and this attribute is only modified by the current task, then a critical section is not required.

Reply:
> ![image](https://private-user-images.githubusercontent.com/758493/389535235-9ba91a6e-6828-4fef-ac03-ccb8881f449b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDA4NzAsIm5iZiI6MTczMjU0MDU3MCwicGF0aCI6Ii83NTg0OTMvMzg5NTM1MjM1LTliYTkxYTZlLTY4MjgtNGZlZi1hYzAzLWNjYjg4ODFmNDQ5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQxMzE2MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOGZiZTk3ODM3N2YyNTgzNjc0ZGI3NzUwZmMzZDIzMGM1NDA4NmI0YjNlYTI4NWU3NGYxNmJlMjhmZTgxMTFiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2MaNol0SIb4K-R9QOQVXGXiZ49-O-dHve0HEhpTitIw) If an ISR is trigger caused context switch when `lockcount == 0` before capture trace, you will see a nested sched_lock() issue will occur in another task from trace point, because the previous `sched_unlock()` is not completed.

It has been updated. I mistakenly removed the interrupt check. Perhaps a version of sched_lock without interrupt checking should be created separately
![image](https://github.com/user-attachments/assets/26cdf38c-bc82-489a-88ce-03be7c44c0df)

Interrupts have been disabled here.

Reply:
Why sched_lock()/sched_unlock() must be coupled with spinlock? They can also be used without disabling interrupts

Reply:
I have added an interrupt check in sched_lock. Will this address your concerns?

Reply:
let us hold csection before decrease lockcount if trace is enabled

Reply:
ok